library/init/meta/coinductive_predicates.lean

```diff
@@ -488,7 +488,8 @@ meta def add_coinductive_predicate
         let (bs, eqs) := compact_relation r.loc_args $ pd.locals.zip r.insts,
         eqs ← eqs.mmap (λ⟨l, i⟩, do
           sort u ← infer_type l.local_type,
-          return $ (const `eq [u] : expr) l.local_type i l),
+          let final : expr tt := (const `eq [u] : expr) l.local_type i l,
```

Comment:
Issue 2: Despite the relevant `has_coe_to_fun` instance being found and in principle completely reducible, it's only with this type annotation that the type of this expression actually gets reduced. Why?